### ISSUE TITLE
Updating index.html

### ISSUE THREAD

**Issue by JakeInit** [CrowCpp/crow#158]
I am looking to send an image to my server. I start by sending the index.html file so the web page reloads on a timer and triggers another GET request:

main.cpp:
```cpp
#define CROW_MAIN
#include "crow_all.h"

int main(int argc, char** argv) {
  crow::SimpleApp app;
  crow::mustache::set_base(".");

  CROW_ROUTE(app, "/image").methods("GET"_method) ([](crow::response& res) {
    res.set_header("Content-type", "text/html");
    res.set_static_file_info("index.html");
    res.end();
  });

  app.loglevel(crow::LogLevel::WARNING);
  app.port(8080).multithreaded().run();
  return 0;
}
```

index.html: 
```html
<!DOCTYPE html>
<html>

	<head>
		<script type="text/javascript">
			function imageToggle()
			{
				date = new Date();
				// Remove old pic
				pic = document.getElementById('img');
				pic.remove();
				// create new pic
				let img = document.createElement("img");
				img.setAttribute("id", "img");
				img.setAttribute("src", "image.png" + "?" + date.getTime());	// Appends date to force reload
				img.setAttribute("class", "responsive");
				document.body.appendChild(img);
			}

			function runTimer()
			{
				setInterval(imageToggle, 1523);
			}

			var pic;
			var date;
		</script>
	<style>
	.responsive{
		width: auto;
  		height: auto;
	}
	</style>
	</head>
	
	<body onload = "runTimer()">
	
		Below is the image<br>
		
		<img id="img" src="image.png" class="responsive" >
	</body>
</html>
```

The index.html is sent and I see GET request at regular intervals. Of course, now I need to be able to get the new image to the web page. I had tried passing in a boolean by reference into the lambda function to control what is sent. I send the index.html on the first iteration and then after send the updated image from then on, but the web page just gets overwritten by the image and I no longer get the interval reload for a GET request. The image changes in my application and I would like it to be shown as the application runs. Is this a functionality that crow can handle or should I start looking into alternate solutions?

**Comment by The-EDev** [comment]
If I understand correctly, you're trying to upload an image to the server and have it show up in your index.html, is that correct?

**Comment by JakeInit** [comment]
This is correct. index.html does not see the image.

**Comment by The-EDev** [comment]
> This is correct. index.html does not see the image.

This is because Crow does not allow access to the server's filesystem. If you want crow to serve a specific file you'll need to place it in a folder named `static` and change your image's src to `static/image.png`. This will make Crow serve the image.

As for uploading, I'm not sure how you're uploading the image, normally you'd need a post request to the server and a handler to receive the file and place it where needed.

Another thing to note is that you're incorrectly serving the html page, while it is possible to use `set_static_file_info()` for html files. It is best to avoid that and instead use `crow::mustache::load("index.html").render()`. The second method is tailor made for html files and also allows you to mustache to add variables to your html page.

For more information on html and static files take a look at [this guide](https://crowcpp.org/guides/static/) and [this guide](https://crowcpp.org/guides/templating/)

**Comment by The-EDev** [comment]
Closing this issue for lack of activity, @JakeInit feel free to reopen if the problem still persists

**Comment by JakeInit** [comment]
Sorry for the disappearance. Is the intention of mustache to replace the html file, handle the html file, either or? The links provided show what appears to be lots of data handling. It looks like I can create templates for the webpage and replace values dependent on what I place in crow::mustache::context. Everything seems dependent on just text or a value passed using the web address.

On opening the address "http://localhost:8080/image" I would use a GET handler as shown below to send the index.html. I have the updated code below based on your comments.

```
static bool sentHtml = false;
CROW_ROUTE(app, "/image").methods("GET"_method) ([&sentHtml]() {
    crow::mustache::context ctx;
    if(!sentHtml) {
      sentHtml = true;
      return crow::mustache::load("index.html").render();
    }
    return crow::mustache::load("image.png").render();
  });
```

1. The above sends the index.html page on the first get request, when the web page first loads, in order to set it up. This works as expected. 
2. The index.html force reloads the page by calling the imageToggle() function each time the interval times out. 
3. This should trigger another get request to the server. 
4. Then the second request should upload the image and any following requests (this does not happen). 
5. This way as the image updates on the server machine, the update can be seen on the web page at the specified interval rate.

The index.html takes the image name and appends a new data time to the string name. This is done since a GET request will cache the image name. I realize that a POST would be better, since it does not cache the data. I'm not certain how to implement this immediately, but realize that is not necessarily a Crow issue. I'll check more into this.

My question, does the algorithm above seem doable using Crow? Can I have a page setup so that I get continuous image updates at set intervals? Maybe I need to write some middle ware for this to work?

 Also, I have been able to modify the above so that I just send the image (simply removing the logic and always sending image.png on each page reload). I have not had to place index.html or image.png in any specific folders for Crow to find them. That may be a setting in the API from what I was reading in the guide.

**Comment by JakeInit** [comment]
I'd like to mention I have gotten this working using python. However, I really need this to be a c++ application. Therefore I'd like to figure out how to get his working. With the below Python code, I can achieve what I have described above. Hopefully positing the python code can give some clarity.

```
#!/usr/bin/python

# from https://www.acmesystems.it/python_http

from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
from os import curdir, sep
import mimetypes
import urlparse
from urlparse import urlparse
from urlparse import parse_qs

class myHandler(BaseHTTPRequestHandler):
	
	#Handler for the GET requests
	def do_GET(self):
		if self.path=="/":
			self.path="/index.html"
		else:
			self.path = "image.png";
			
		mimetypes.add_type('image/x-portable-graymap', '.pgm')
			
		print "Path = " + self.path
		

		try:
			sendReply = False
			if self.path.endswith(".jpg"):
				mimetype='image/jpg'
				sendReply = True
			if self.path.endswith(".jpeg"):
				mimetype='image/jpeg'
				sendReply = True
			if self.path.endswith(".pgm"):
				mimetype='image/x-portable-graymap'
				sendReply = True
			if self.path.endswith(".gif"):
				mimetype='image/gif'
				sendReply = True
			if self.path.endswith(".png"):
				mimetype='image/png'
				sendReply = True
			if self.path.endswith(".html"):
				mimetype='text/html'
				sendReply = True
			if self.path.endswith(".js"):
				mimetype='application/javascript'
				sendReply = True

			if sendReply == True:
				print "\n"
				#Open the static file requested and send it
				f = open(curdir + sep + self.path)
				self.send_response(200)
				self.send_header('Content-type', mimetype)
				self.end_headers()
				self.wfile.write(f.read())
				f.close()
			else:
				print "\n"
			return


		except IOError:
			self.send_error(404,'File Not Found: %s' % self.path)
			
objectHandler = myHandler
PORT_NUMBER = 8080

try:
	#Create a web server and define the handler to manage the
	#incoming request
	server = HTTPServer(('', PORT_NUMBER), objectHandler)
	print 'Started httpserver on port ' , PORT_NUMBER
	
	#Wait forever for incoming http requests
	server.serve_forever()

except KeyboardInterrupt:
	print '\nShutdown signal received, closing the web server'
	server.socket.close()
```
The above pulls the file from the host system. It sends index.html on the first GET request and uploads the image on every following request caused by imageToggle().

Ultimately, I'll be saving image.png in a buffer (maybe base64 encoded). Therefore I'll likely send the buffer using a json and will not necessarily need access to the file system.

**Comment by The-EDev** [comment]
Alright I think I have a clear Idea of what you're trying to do.

First let's clear mustache up, mustache is a templating system, meaning it takes an html file, finds any tags, and replaces them with what it can find in the context provided. It does not modify any files, it is there for you to have one page with different variables. It also does not handle anything other than html files.

jumping down to your question:
> Can I have a page setup so that I get continuous image updates at set intervals?
Absolutely, you don't need middleware either. 

Moving on to your second comment, it looks like you only care whether the request is going to "/" or not. Which is also doable in crow, it's just that you're not using the right tools that crow provides. Give me a some time and I'll rewrite your python code for Crow.

**Comment by The-EDev** [comment]
The C++ code to get your system working is this:
```cpp
#define CROW_MAIN
//#define CROW_STATIC_DIRECTORY ""
//#define CROW_STATIC_ENDPOINT "/<path>"
#include "crow_all.h"

int main(int argc, char** argv) {
  crow::SimpleApp app;
  crow::mustache::set_base(".");

  CROW_ROUTE(app, "/") ([]() 
  {
    return crow::mustache::load("index.html").render();
  });

  CROW_ROUTE(app, "/image.png")([](crow::response& res)
  {
    res.set_static_file_info("image.png");
    res.end();
  });

  app.loglevel(crow::LogLevel::WARNING);
  app.port(8080).multithreaded().run();
  return 0;
}
```
Note: You can use the commented lines and remove the second route, but this will give anyone access to whatever files are in your system, which isn't the best idea.

I modified the html code a bit, but that shouldn't cause any problems:
```js
//OLD
img.setAttribute("src", "image.png" + "?" + date.getTime());	// Appends date to force reload
//NEW
img.setAttribute("src", "image.png?time=" + date.getTime());	// Appends date to force reload
```

Now I don't know how you plan on updating the Image. There isn't a way to upload images from the client to the server, but that's outside the scope of this question and issue.

Let me know if you need any more help.

**Comment by JakeInit** [comment]
The above makes sense. Thank you for your responsiveness and effort to help me use crow properly. The update for the html will be a little more descriptive. As far as updating the image, That is done within my application already. I was looking for a way to visualize that and I believe you have helped. Thanks again.

**Comment by The-EDev** [comment]
No problem, I changed the html because it was sending the time as a key without a value. I'm happy to help with any other questions you might have :)

**Comment by JakeInit** [comment]
I have one last question on accessing the web page. Using the provided code you had written, the application works great. I can navigate to http://localhost:8080 and I'll see the update of the image via the index.html. However, if I try to extend the url, the update of the image does not occur.

```
CROW_ROUTE(app, "/helloworld") ([]() {
    return crow::mustache::load("index.html").render();
  });

  CROW_ROUTE(app, "/helloworld/image.png")([](crow::response& res) {
    res.set_static_file_info("image.png");
    res.end();
  });
```

On navigating to http://localhost:8080/helloworld, the index.html is loaded up just fine. For some reason, I'm not getting the update of the image however.

**Comment by The-EDev** [comment]
> I have one last question on accessing the web page. Using the provided code you had written, the application works great. I can navigate to http://localhost:8080 and I'll see the update of the image via the index.html. However, if I try to extend the url, the update of the image does not occur.
> 
> ```
> CROW_ROUTE(app, "/helloworld") ([]() {
>     return crow::mustache::load("index.html").render();
>   });
> 
>   CROW_ROUTE(app, "/helloworld/image.png")([](crow::response& res) {
>     res.set_static_file_info("image.png");
>     res.end();
>   });
> ```
> 
> On navigating to http://localhost:8080/helloworld, the index.html is loaded up just fine. For some reason, I'm not getting the update of the image however.

This is because while your page is at `"/helloworld"` your script is still making a request to `"/image.png"` as opposed to `"/helloworld/image.png"`.
To solve it, you can either remove the extra part from your route, or add it to the script, making it either `CROW_ROUTE(app, "/image.png")` or `img.setAttribute("src", "helloworld/image.png?time=" + date.getTime());`

**Comment by JakeInit** [comment]
Ok,
That worked. I kept the index.html file as is and removed the extra bit from the route. As an exercise, I also looked into what would happen if I extended the web address even further.

```
CROW_ROUTE(app, "/helloworld/picture") ([]() {
    return crow::mustache::load("index.html").render();
  });

// my intuition tells me to keep the original solution
// CROW_ROUTE(app, "/image.png")([](crow::response& res) {
// but this did not work and had to change to:
  CROW_ROUTE(app, "/helloworld/image.png")([](crow::response& res) {  // This line works
    res.set_static_file_info("image.png");
    res.end();
  });
```

I would have thought with the index.html unchanged, the request would still be for /image.png. It now seems like that the request is always at 1 layer above the page index. So before it was just at localhost:portNumber/image.png when the page was at localhost:portNumber/helloworld and then the call was to localhost:portNumber/helloworld/image.png when the page was at localhost:portNumber/helloworld/picture.

At this point I have it doing everything I need. Thanks again for your help.
